File: Coding/Sorting Algo/countSort/Code.cpp
/*
@ Creator: Amaan Shaikh
@ Problem_Title: countSort
    @ Problem_Statement: 
    @ Data_Structure: 
    @ Algorithm:
    @ Level: 
    @ URL: 
@ Approach: Optimized Approach 1
@ Date_Time: June 22, 2022 07:17:43
*/

#include<bits/stdc++.h>

using namespace std;

void countSort(int* arr, int n){
    int K = INT_MIN;
    for(int i = 0; i < n; i++){
        K = max(arr[i],K);
    }
    K++;
    int counter[K];
    for(int i = 0; i < K; i++){
        counter[i] = 0;
    }

    for(int i = 0; i < n; i++){
        counter[arr[i]]++;
    }

    for(int i = 1; i < K; i++){
        counter[i]+=counter[i-1];
    }
    
    int result[n];
    for(int i = n-1; i >= 0; i--){
        result[--counter[arr[i]]] = arr[i];
    }

    for(int i = 0; i < n; i++){
        arr[i] = result[i];
    }
}

int main(){
    int n = 0;
    cin >> n;
    int arr[n];
    for(int i = 0; i < n; i++){
        cin >> arr[i];
    }
    countSort(arr, n);
    for(int i = 0; i < n; i++){
        cout << arr[i] << " ";
    }cout << endl;
    return 0;
}

File: Coding/Deque/IntroToDQ/Code.java
/*
@ Creator: Amaan Shaikh
@ Problem_Title: IntroToDQ
    @ Problem_Statement: 
    @ Data_Structure: 
    @ Algorithm:
    @ Level: 
    @ URL: 
@ Approach: Introduction to Deque
@ Date_Time: August 19, 2022 14:14:42
*/

// import java.util.*;
import java.util.ArrayDeque;
import java.util.Iterator;
import java.util.TreeSet;

class Code{
    public static void main(String[] args){
        ArrayDeque<Integer> dq = new ArrayDeque<>();
        dq.offerFirst(1);
        dq.offerFirst(2);
        dq.offerFirst(3);
        dq.offerFirst(4);
        dq.offerFirst(5);

        dq.offerLast(6);
        dq.offerLast(7);
        dq.offerLast(8);
        dq.offerLast(9);
        System.out.println(dq);

        System.out.println(dq.pollFirst());
        System.out.println(dq.pollLast());
        System.out.println(dq.peek());
        System.out.println(dq.peekLast());
        System.out.println(dq);
        Iterator<Integer> it = dq.iterator();
        for(Integer ele : dq){
            System.out.print(ele+" _ ");
        }
        System.out.println();
        while(it.hasNext()){
            System.out.println(it.next());
        }
        

    } 
}